File: Problems/Data-structures/Day-07/sol/ayush2005k/Solution1.py
"""
Problem: C. Number of Pairs
Link: https://codeforces.com/problemset/problem/1538/C

Short Problem Statement:
Given an array of integers, count the number of index pairs (i, j)
such that i < j and the sum a[i] + a[j] lies between l and r (inclusive).

Approach:
- Sort the array.
- For each index i, we want to find how many indices j > i satisfy:
    l - a[i] <= a[j] <= r - a[i]
- Since the array is sorted, binary search can be used to find
  the valid range of j efficiently.
- Sum the counts for all i.

Time Complexity:
O(n log n) per test case

Space Complexity:
O(n)

Example:
Input:
3 4 7
5 1 2

Output:
2

Submission Link:
https://codeforces.com/contest/1538/submission/356210806
"""

import sys
from bisect import bisect_left, bisect_right

input = sys.stdin.readline

t = int(input())

for _ in range(t):
    n, l, r = map(int, input().split())
    a = list(map(int, input().split()))

    a.sort()
    ans = 0

    for i in range(n):
        left = bisect_left(a, l - a[i], i + 1)
        right = bisect_right(a, r - a[i], i + 1)
        ans += (right - left)

    print(ans)

File: Problems/Binary-Search/Day-07/sol/Solution1.cpp
//link : https://codeforces.com/problemset/submission/1538/356089286


#include <bits/stdc++.h>
using namespace std;

#define fastio ios::sync_with_stdio(false); cin.tie(NULL);

using ll = long long;
using vi = vector<ll>;

void solve() {
    int n;
    ll l, r;
    cin >> n >> l >> r;

    vi a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    sort(a.begin(), a.end());

    ll ans = 0;

    for (int i = 0; i < n; i++) {
        ll low = l - a[i];
        ll high = r - a[i];

        int left = lower_bound(a.begin() + i + 1, a.end(), low) - a.begin();
        int right = upper_bound(a.begin() + i + 1, a.end(), high) - a.begin();

        ans += max(0, right - left);
    }

    cout << ans << '\n';
}

int main() {
    fastio

    int t;
    cin >> t;
    while (t--) {
        solve();
    }

    return 0;
}



/*

Explaination:
1. Sort the array.
2. For each index i, we want to count indices j > i such that:
      l ≤ a[i] + a[j] ≤ r
   This can be rewritten as:
      l - a[i] ≤ a[j] ≤ r - a[i]
3. Since the array is sorted, use binary search:
   - lower_bound to find the first j where a[j] ≥ (l - a[i])
   - upper_bound to find the first j where a[j] > (r - a[i])
4. The number of valid pairs for index i is (right - left).
5. Sum this for all i.

Complexity:
- Time complexity: O(n log n)
- Space complexity: O(n)

*/